devU-api/src/containerAutoGrader/containerAutoGrader.controller.ts

```diff
@@ -0,0 +1,102 @@
+import { Request, Response, NextFunction } from 'express'
+
+import ContainerAutoGraderService from './containerAutoGrader.service'
+import { serialize } from './containerAutoGrader.serializer'
+
+import { GenericResponse, NotFound, Updated } from '../utils/apiResponse.utils'
+
+export async function get(req: Request, res: Response, next: NextFunction) {
+  try {
+    const containerAutoGrader = await ContainerAutoGraderService.list()
+
+    res.status(200).json(containerAutoGrader.map(serialize))
+  } catch (err) {
+    next(err)
+  }
+}
+
+export async function detail(req: Request, res: Response, next: NextFunction) {
+  try {
+    const id = parseInt(req.params.id)
+    const containerAutoGrader = await ContainerAutoGraderService.retrieve(id)
+
+    if (!containerAutoGrader) return res.status(404).json(NotFound)
+
+    const response = serialize(containerAutoGrader)
+
+    res.status(200).json(response)
+  } catch (err) {
+    next(err)
+  }
+}
+
+/*
+  * for the post method, I changed how the upload is handled. I am now using fields instead of
+  * single(for the purpose of uploading grader file and makefile). But I set the makefile to be
+  * optional, since it is not required. I also added the makefile to the create method in the
+  * ContainerAutoGraderService.
+*/
+export async function post(req: Request, res: Response, next: NextFunction) {
+  try {
+    if (!req.files || !('graderFilename' in req.files)) {
+      return res.status(400).json(new GenericResponse('Container Auto Grader requires file upload for grader'));
+    }
+    const graderFile = req.files['graderFilename'][0]?.buffer
+    const makefile = req.files['makefileFilename'] ? req.files['makefileFilename'][0]?.buffer : null
```

Comment:
These are not filenames at this point, they are the actual contents of the file. Then only the filenames are stored in the DB. Please change the field names to make this clear

Reply:
modified via commit id [a05bf00] with also type in types  and all of related files